### PR TITLE
Fix permission issue in `get_nfs_cache_root_dir`

### PR DESCRIPTION
**Pull request by WeichenXu123**
Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Fix permission issue in `get_nfs_cache_root_dir`

On databricks, if running on cluster with user isolation,
non-root user cannot access `/local_disk0/.ephemeral_nfs` directory, that will cause `get_nfs_cache_root_dir` return False and cause NFS related optimization disabled.
We can try to call `getReplNFSTempDir` first, the directory `getReplNFSTempDir` returned is always readable and writable.

## How is this patch tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [x] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/nfs_on_spark.py [hunk 1]
```diff
@@ -27,18 +28,23 @@ def get_nfs_cache_root_dir():
...
         )
         if nfs_enabled:
-            nfs_root_dir = "/local_disk0/.ephemeral_nfs"
-            # Test whether the NFS directory is writable.
-            test_path = os.path.join(nfs_root_dir, uuid.uuid4().hex)
             try:
-                os.makedirs(test_path)
-                return nfs_root_dir
+                # If databricks runtime has dbutils entrypoint function `getReplNFSTempDir`,
+                # the directory it returns always has available permission.
```
Comment (harupy) [inline mlflow/utils/nfs_on_spark.py:33]: What is "available permission"?

Comment (WeichenXu123) [reply mlflow/utils/nfs_on_spark.py:33]: The returned directory is ensured to be read / write

Comment (harupy) [reply mlflow/utils/nfs_on_spark.py:33]: If so, please say so.

File: mlflow/utils/nfs_on_spark.py [hunk 2]
```diff
@@ -27,18 +28,23 @@ def get_nfs_cache_root_dir():
...
-                shutil.rmtree(test_path, ignore_errors=True)
+                nfs_root_dir = "/local_disk0/.ephemeral_nfs"
+                # Test whether the NFS directory is writable.
+                test_path = os.path.join(nfs_root_dir, uuid.uuid4().hex)
+                try:
+                    os.makedirs(test_path)
+                    return nfs_root_dir
+                except Exception:
+                    # For databricks cluster enabled Table ACL, we have no permission to access NFS
+                    # directory, in this case, return None representing NFS is not available.
```
Comment (harupy) [inline mlflow/utils/nfs_on_spark.py:44]:
```suggestion
                    # directory, in this case, return None, meaning NFS is not available.
```

File: mlflow/utils/nfs_on_spark.py [hunk 3]
```diff
@@ -27,18 +28,23 @@ def get_nfs_cache_root_dir():
             spark_sess.conf.get("spark.databricks.mlflow.nfs.enabled", "true").lower() == "true"
         )
         if nfs_enabled:
-            nfs_root_dir = "/local_disk0/.ephemeral_nfs"
-            # Test whether the NFS directory is writable.
-            test_path = os.path.join(nfs_root_dir, uuid.uuid4().hex)
             try:
-                os.makedirs(test_path)
-                return nfs_root_dir
+                # If databricks runtime has dbutils entrypoint function `getReplNFSTempDir`,
```
Comment (harupy) [inline mlflow/utils/nfs_on_spark.py:32]:
```suggestion
```

I don't think we need this line.

File: mlflow/utils/nfs_on_spark.py [hunk 4]
```diff
@@ -27,18 +28,23 @@ def get_nfs_cache_root_dir():
...
+                # the directory it returns is ensured to have read / write permission granted.
```
Comment (harupy) [inline mlflow/utils/nfs_on_spark.py:33]:
```suggestion
                # the directory `getReplNFSTempDir` returns has read/write permissions.
```

**Review by harupy — APPROVED** [review]

**Pull request merged**